### PR TITLE
[14.0][MIG+REF+IMP] github_connector: Migration to 14.0, refactor and improvements

### PR DESCRIPTION
**Pull request by joao-p-marques**
- Remove useless (and wrong) `ondelete` attribute from O2M field in `github.organization`. Odoo O2M fields don't support it.
- Rename and convert relation between `res.partner` and `github.team.partner`. The latest will be unique for each user in each organization, so it makes no sense to make it a M2M, but rather a O2M. Also, the relation is to the "Team Partner" entity, and not to "Team".
- Remove username/password auth. Deprecated by Github: https://docs.github.com/en/rest/overview/other-authentication-methods#via-username-and-password
- Allow to set the Github acces token as an IR Configuration Parameter. Whatever is set in the config file still takes priority.
- Remove useless demo data
- Move to [PyGithub](https://pygithub.readthedocs.io/en/latest/)
- Refactor tests to be independent of a real connection to Github. Use https://github.com/getsentry/responses to mock calls to the API, and ensure a consistent behavior without the need of setting up secrets and a custom connection. The real connection is now handled by `PyGithub` and is version pinned as to the API spec.

@Tecnativa
TT29979

ping @pedrobaeza @victoralmau

### CODE REVIEW

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Thanks for migrating module and all the changes. 
Just one thing regarding "Remove useless demo data". I don't see it is useless, as it allow you to test very easely a module. So please keep demo data as it doesn't hurt.

thanks !

**Comment by joao-p-marques** [comment]
Thanks for the review @legalsylvain :+1: 

> Just one thing regarding "Remove useless demo data". I don't see it is useless, as it allow you to test very easely a module. So please keep demo data as it doesn't hurt.

Sorry, maybe "Remove useless demo data" has been a little uninformative and harsh :sweat_smile: I didn't remove all the data, just the demo modules. That is good to test, but at the end those don't map real modules in Github, so you were always unable to test the full flow. IMHO it would make sense to just keep as demo data the one that is "real" and that will allow you to perform the full workflow of the module... With this change, I kept only the OCA Github organization, which then allows the user to quickly sync the respective repositories and users and test the module correctly. 

Another option would be to add a couple of demo modules but that referenced real modules in Github. However, as the OCA changes, that can become problematic and not be worth the effort...

File: github_connector/models/github.py [hunk 1]
```diff
@@ -1,160 +0,0 @@
-# Copyright (C) 2016-Today: Odoo Community Association (OCA)
```
Comment (legalsylvain) [inline github_connector/models/github.py]: nice simplification ! thanks !

**Comment by legalsylvain** [comment]
you should add a 'github.access_token' configuration parameter to fix travis.

**Comment by legalsylvain** [comment]
> Sorry, maybe "Remove useless demo data" has been a little uninformative and harsh sweat_smile I didn't remove all the data, just the demo modules. That is good to test, but at the end those don't map real modules in Github, so you were always unable to test the full flow. IMHO it would make sense to just keep as demo data the one that is "real" and that will allow you to perform the full workflow of the module... With this change, I kept only the OCA Github organization, which then allows the user to quickly sync the respective repositories and users and test the module correctly.

:+1: . sorry, I just read your description, and not the full diff. thanks for the explanation. (And thanks for porting this module !)

**Comment by joao-p-marques** [comment]
> > Sorry, maybe "Remove useless demo data" has been a little uninformative and harsh sweat_smile I didn't remove all the data, just the demo modules. That is good to test, but at the end those don't map real modules in Github, so you were always unable to test the full flow. IMHO it would make sense to just keep as demo data the one that is "real" and that will allow you to perform the full workflow of the module... With this change, I kept only the OCA Github organization, which then allows the user to quickly sync the respective repositories and users and test the module correctly.
> 
> +1 . sorry, I just read your description, and not the full diff. thanks for the explanation. (And thanks for porting this module !)

No problem, thanks for the review :+1: I think all is good now and tests are green, thanks to the little refactor (I added an extra point in the original description)

**Review by victoralmau — APPROVED** [review]
Code review OK.

File: github_connector/views/view_github_team_partner.xml [hunk 2]
```diff
@@ -21,11 +21,11 @@ License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
             <tree decoration-warning="role == 'member'">
                 <field
                     name="team_id"
-                    invisible="context.get('search_default_team_id')"
+                    attrs="{'invisible': 'context.get(\'search_default_team_id\')'}"
```
Comment (pedrobaeza) [inline github_connector/views/view_github_team_partner.xml]: Incorrect expression:
```suggestion
                    attrs="{'invisible': context.get('search_default_team_id')}"
```
and check the rest.

Comment (joao-p-marques) [reply github_connector/views/view_github_team_partner.xml]: If I do that, I get this error:

```
Odoo Server Error

Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 237, in _dispatch
    result = request.dispatch()
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 683, in dispatch
    result = self._call_function(**self.params)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 359, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 347, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 912, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 531, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/auto/addons/web/controllers/main.py", line 1389, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/auto/addons/web/controllers/main.py", line 1381, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/api.py", line 392, in call_kw
    result = _call_kw_model(method, model, args, kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/api.py", line 365, in _call_kw_model
    result = method(recs, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 1527, in load_views
    result['fields_views'] = {
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 1528, in <dictcomp>
    v_type: self.fields_view_get(v_id, v_type if v_type != 'list' else 'tree',
  File "/opt/odoo/auto/addons/web/models/models.py", line 220, in fields_view_get
    r = super().fields_view_get(view_id, view_type, toolbar, submenu)
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 1618, in fields_view_get
    xarch, xfields = view.postprocess_and_fields(etree.fromstring(result['arch']), model=self._name)
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_ui_view.py", line 851, in postprocess_and_fields
    arch, name_manager = self._postprocess_view(node, model, validate=validate)
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_ui_view.py", line 863, in _postprocess_view
    self.postprocess(node, [], editable, name_manager)
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_ui_view.py", line 963, in postprocess
    self.postprocess(child, current_node_path, node_info['editable'], name_manager)
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_ui_view.py", line 958, in postprocess
    transfer_node_to_modifiers(node, node_info['modifiers'], self._context, current_node_path)
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_ui_view.py", line 66, in transfer_node_to_modifiers
    modifiers.update(ast.literal_eval(attrs))
  File "/usr/local/lib/python3.8/ast.py", line 99, in literal_eval
    return _convert(node_or_string)
  File "/usr/local/lib/python3.8/ast.py", line 88, in _convert
    return dict(zip(map(_convert, node.keys),
  File "/usr/local/lib/python3.8/ast.py", line 98, in _convert
    return _convert_signed_num(node)
  File "/usr/local/lib/python3.8/ast.py", line 75, in _convert_signed_num
    return _convert_num(node)
  File "/usr/local/lib/python3.8/ast.py", line 66, in _convert_num
    _raise_malformed_node(node)
  File "/usr/local/lib/python3.8/ast.py", line 63, in _raise_malformed_node
    raise ValueError(f'malformed node or string: {node!r}')
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 639, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 315, in _handle_exception
    raise exception.with_traceback(None) from new_cause
ValueError: malformed node or string: <_ast.Call object at 0x7f32a2c8a520>
```

Everything works with it as is. The truth is I don't recall why I needed this change :thinking: but it seems necessary, right? Or am I doing anything wrong somewhere else?

Comment (pedrobaeza) [reply github_connector/views/view_github_team_partner.xml]: The question is that if your expressing is working to show conditionally things. If you quote the whole expression as a string, I doubt it works as intended.

Comment (joao-p-marques) [reply github_connector/views/view_github_team_partner.xml]: Yes, you are right. I already added the correct fix to make it work correctly :+1:

File: github_connector/views/view_github_team_partner.xml [hunk 3]
```diff
@@ -21,13 +21,15 @@ License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
             <tree decoration-warning="role == 'member'">
                 <field
                     name="team_id"
-                    invisible="context.get('search_default_team_id')"
```
Comment (pedrobaeza) [inline github_connector/views/view_github_team_partner.xml]: I don't get why this is not working.

Comment (joao-p-marques) [reply github_connector/views/view_github_team_partner.xml]: AFAICS, it is due to the removal of dynamic expressions for the `invisible` and `readonly` attributes in XML for v14. We are forced to use `attrs` instead, and declare it there, and for some reason it is not accepting the syntax you suggested in https://github.com/OCA/interface-github/pull/73#discussion_r668002750

I dig a little and found that this is what we already had to do in https://github.com/OCA/product-attribute/pull/742/commits/78b534eced1166f77525b3de1575df369533b038#diff-68f52c9b5cbb5cad326ecb91bf58f75aa8b5a47d4b82f88b0e1dd3b575ea6df7L11-L20

Comment (pedrobaeza) [reply github_connector/views/view_github_team_partner.xml]: But I still see this expression on v14 code: https://github.com/odoo/odoo/blob/14.0/addons/mail/wizard/invite_view.xml#L17

Comment (joao-p-marques) [reply github_connector/views/view_github_team_partner.xml]: Mmm yes, then is somehting wrong with the migration steps? https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-14.0#tasks-to-do-in-the-migration

> In View XML, the `invisible` and `readonly` no longer support dynamic expressions. When using those, use the `attrs` attribute instead

Comment (pedrobaeza) [reply github_connector/views/view_github_team_partner.xml]: I have changed such line to:
```
* In the XML views, the invisible and readonly no longer support dynamic 
  expressions like `state=='draft'`. When using those, use the `attrs` attribute 
  instead. You can only use them for direct context queries like `context.get('invisible')`.
```

File: github_connector/models/github_team_partner.py [hunk 4]
```diff
@@ -34,3 +34,22 @@ class GithubTeamPartner(models.Model):
...
     )
+
+    context_search_default_team_id = fields.Integer(
+        compute="_compute_context_search_default"
+    )
+    context_search_default_partner_id = fields.Integer(
+        compute="_compute_context_search_default"
+    )
+
+    def _compute_context_search_default(self):
```
Comment (ernestotejeda) [inline github_connector/models/github_team_partner.py:46]: Don't need to decorate these (context dependent) methods with @api.depends_context? In this case @api.depends_context('search_default_team_id', 'search_default_partner_id')
ping @joao-p-marques @pedrobaeza

Comment (pedrobaeza) [reply github_connector/models/github_team_partner.py:46]: Indeed.

**Comment by ernestotejeda** [comment]
@pedrobaeza @joao-p-marques : I added ``@api.depends_context`` decorator in the methods that need it

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**